File: data-access-layer/modules/compliance/svc.ts
import { redisClient, type RedisClientType } from "../../db/redis-client.ts";
import {
  type PropertyParentDocument,
  REQUEST_STATUS,
  type RequestStatus,
} from "./model.ts";
import { ComplianceRepo, type IComplianceRepo } from "./repo.ts";
import { pool } from "../../db/db.ts";
import type { PagedResponse } from "../../../types/pagination.ts";
import { COMPLIANCE_EVENTS } from "../../shared/events/compliance.ts";
import { createEvent } from "../../shared/types/event.ts";
import type { Server as SocketServer } from "socket.io";
import { getSocketServer } from "../../../socketServer.ts";

export interface IComplianceSvc {
  getAllDocumentApprovalRequestsForGivenProperty(
    propertyId: string,
    pageSize: number,
    pageNumber: number,
  ): Promise<{ data?: PagedResponse<PropertyParentDocument>; error?: Error }>;
  getAllDocumentApprovalRequestsForGivenPropertyParent(
    propertyId: string,
    userId: string,
  ): Promise<{ data?: PropertyParentDocument[]; error?: Error }>;
  getPropertyParentDocumentApprovalRequestByDocumentId(
    propertyId: string,
    userId: string,
    parentDocId: string,
  ): Promise<{ data?: PropertyParentDocument; error?: Error }>;
  sendPropertyParentDocumentApprovalRequest(
    propertyId: string,
    userId: string,
    parentDocumentId: string,
  ): Promise<{ data?: PropertyParentDocument; error?: Error }>;
  setPropertyParentDocumentRequestStatus(
    propertyId: string,
    userId: string,
    parentDocumentId: string,
    requestStatus: RequestStatus,
    adminId: string,
  ): Promise<{ data?: PropertyParentDocument; error?: Error }>;
  isPropertyParentDocumentRequestApproved(
    propertyId: string,
    userId: string,
    parentDocumentId: string,
  ): Promise<{ data?: boolean; error?: Error }>;
}

class ComplianceSvc implements IComplianceSvc {
  private redisClient: RedisClientType;
  private socketServer: SocketServer | null = null;
  private repo: IComplianceRepo;
  constructor(
    redisClient: RedisClientType,
    socketServer: SocketServer | null = null,
    repo?: IComplianceRepo,
  ) {
    this.redisClient = redisClient;
    this.socketServer = socketServer;
    this.repo = repo ?? new ComplianceRepo(pool, redisClient);
  }

  async getAllDocumentApprovalRequestsForGivenPropertyParent(
    propertyId: string,
    userId: string,
  ): Promise<{ data?: PropertyParentDocument[]; error?: Error }> {
    return this.repo.getAllDocumentApprovalRequestsForGivenPropertyParent(
      propertyId,
      userId,
    );
  }

  async getPropertyParentDocumentApprovalRequestByDocumentId(
    propertyId: string,
    userId: string,
    parentDocId: string,
  ): Promise<{ data?: PropertyParentDocument; error?: Error }> {
    return this.repo.getPropertyParentDocumentApprovalRequestByDocumentId(
      propertyId,
      userId,
      parentDocId,
    );
  }

  async sendPropertyParentDocumentApprovalRequest(
    propertyId: string,
    userId: string,
    parentDocumentId: string,
  ): Promise<{ data?: PropertyParentDocument; error?: Error }> {
    const { data, error } =
      await this.repo.sendPropertyParentDocumentApprovalRequest(
        propertyId,
        userId,
        parentDocumentId,
      );
    if (error) return { data: undefined, error: error };

    if (!this.socketServer) this.initSocketServer();
    this.socketServer?.emit(
      COMPLIANCE_EVENTS.PROPERTY_PARENT_DOCUMENT_REQUESTED,
      data,
    );

    return { data, error: undefined };
  }

  async getAllDocumentApprovalRequestsForGivenProperty(
    propertyId: string,
    pageSize: number,
    pageNumber: number,
  ): Promise<{ data?: PagedResponse<PropertyParentDocument>; error?: Error }> {
    return this.repo.getAllDocumentApprovalRequestsForGivenProperty(
      propertyId,
      pageSize,
      pageNumber,
    );
  }

  async setPropertyParentDocumentRequestStatus(
    propertyId: string,
    userId: string,
    parentDocumentId: string,
    requestStatus: RequestStatus,
    adminId: string,
  ): Promise<{ data?: PropertyParentDocument; error?: Error }> {
    const { data, error } =
      await this.repo.setPropertyParentDocumentRequestStatus(
        propertyId,
        userId,
        parentDocumentId,
        requestStatus,
        adminId,
      );
    if (error) return { data: undefined, error: error };
    switch (requestStatus) {
      case REQUEST_STATUS.APPROVED:
        console.log("Emitting event...");
        this.redisClient.publish(
          COMPLIANCE_EVENTS.PROPERTY_PARENT_DOCUMENT_APPROVED,
          JSON.stringify(
            createEvent(
              COMPLIANCE_EVENTS.PROPERTY_PARENT_DOCUMENT_APPROVED,
              "compliance",
              "1.0",
              data,
            ),
          ),
        );
      //TODO: Rejected
    }
    return { data: data, error: undefined };
  }

  async isPropertyParentDocumentRequestApproved(
    propertyId: string,
    userId: string,
    parentDocumentId: string,
  ): Promise<{ data?: boolean; error?: Error }> {
    return await this.repo.isPropertyParentDocumentRequestApproved(
      propertyId,
      userId,
      parentDocumentId,
    );
  }

  private initSocketServer() {
    const socketServer = getSocketServer();
    if (!socketServer) return;
    this.socketServer = socketServer;
  }
}

const complianceSvc = new ComplianceSvc(redisClient);
export default complianceSvc;

File: apps/backend-python/src/modules/compliance/svc.py
from abc import ABC, abstractmethod
from typing import List
from uuid import UUID

from src.modules.compliance.client import IIdentityClient
from src.modules.compliance.repo import IComplianceRepo
from src.shared.types.modules.compliance.enum import REQUEST_STATUS
from src.shared.types.modules.compliance.model import (
    PropertyChildDocument,
    PropertyParentDocument,
)
from src.shared.types.pagination import PagedResponse
from src.shared.types.response import HTTPError, HTTPErrorResponse


class IComplianceSvc(ABC):
    @abstractmethod
    async def get_all_document_approval_requests_for_given_property_parent(
        self,
        property_id: UUID,
        user_id: UUID,
    ) -> HTTPErrorResponse[List[PropertyParentDocument]]:
        pass

    @abstractmethod
    async def get_all_document_approval_requests_for_given_property_child(
        self,
        property_id: UUID,
        child_id: UUID,
    ) -> HTTPErrorResponse[List[PropertyChildDocument]]:
        pass

    @abstractmethod
    async def get_property_parent_document_approval_request_by_document_id(
        self,
        property_id: UUID,
        user_id: UUID,
        parent_doc_id: UUID,
    ) -> HTTPErrorResponse[PropertyParentDocument]:
        pass

    @abstractmethod
    async def get_all_document_approval_requests_for_given_property(
        self,
        property_id: UUID,
        page_size: int,
        page_number: int,
    ) -> HTTPErrorResponse[PagedResponse[PropertyParentDocument]]:
        pass

    @abstractmethod
    async def send_property_parent_document_approval_request(
        self,
        property_id: UUID,
        user_id: UUID,
        parent_document_id: UUID,
    ) -> HTTPErrorResponse[PropertyParentDocument]:
        pass

    @abstractmethod
    async def set_property_parent_document_request_status(
        self,
        property_id: UUID,
        user_id: UUID,
        parent_document_id: UUID,
        request_status: REQUEST_STATUS,
        admin_id: UUID,
    ) -> HTTPErrorResponse[PropertyParentDocument]:
        pass


class ComplianceSvc(IComplianceSvc):
    def __init__(self, repo: IComplianceRepo, identity_client: IIdentityClient) -> None:
        self.repo = repo
        self.identity_client = identity_client

    async def get_all_document_approval_requests_for_given_property_parent(
        self,
        property_id: UUID,
        user_id: UUID,
    ) -> HTTPErrorResponse[List[PropertyParentDocument]]:
        return await self.repo.get_all_document_approval_requests_for_given_property_parent(
            property_id=property_id, user_id=user_id
        )

    async def get_all_document_approval_requests_for_given_property_child(
        self,
        property_id: UUID,
        child_id: UUID,
    ) -> HTTPErrorResponse[List[PropertyChildDocument]]:
        return await self.repo.get_all_document_approval_requests_for_given_property_parent(
            property_id=property_id, user_id=child_id
        )

    async def get_property_parent_document_approval_request_by_document_id(
        self,
        property_id: UUID,
        user_id: UUID,
        parent_doc_id: UUID,
    ) -> HTTPErrorResponse[PropertyParentDocument]:
        return await self.repo.get_property_parent_document_approval_request_by_document_id(
            property_id=property_id, user_id=user_id, parent_doc_id=parent_doc_id
        )

    async def get_all_document_approval_requests_for_given_property(
        self,
        property_id: UUID,
        page_size: int,
        page_number: int,
    ) -> HTTPErrorResponse[PagedResponse[PropertyParentDocument]]:
        return await self.repo.get_all_document_approval_requests_for_given_property(
            property_id=property_id,
            page_size=page_size,
            page_number=page_number,
        )

    async def send_property_parent_document_approval_request(
        self,
        property_id: UUID,
        user_id: UUID,
        parent_document_id: UUID,
    ) -> HTTPErrorResponse[PropertyParentDocument]:
        return await self.repo.send_property_parent_document_approval_request(
            property_id=property_id,
            user_id=user_id,
            parent_document_id=parent_document_id,
        )

    async def set_property_parent_document_request_status(
        self,
        property_id: UUID,
        user_id: UUID,
        parent_document_id: UUID,
        request_status: REQUEST_STATUS,
        admin_id: UUID,
    ) -> HTTPErrorResponse[PropertyParentDocument]:
        is_admin, error = await self.identity_client.is_property_admin(
            property_id=property_id, user_id=admin_id
        )
        if error:
            return None, error
        if not is_admin:
            return None, HTTPError(
                code=403, message="Insufficient permissions - required role: 'admin'!"
            )
        data, error = await self.repo.set_property_parent_document_request_status(
            property_id,
            user_id,
            parent_document_id,
            request_status,
            admin_id,
        )
        if error:
            return None, error
        assert data is not None
        # TODO: Publish event
        return data, None
